perf/xtm.cc

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "xtm/xtm_api.h"
+#include "../test/unit.h"
+/** Conflict name with same macro in benchmark.h */
+#undef is
+
+#include <pthread.h>
+
+#include <benchmark/benchmark.h>
+
+enum {
+	/** Maximum message count in xtm queue. */
+	XTM_TEST_QUEUE_SIZE = 64 * 1024,
+	/**
+	 * Maximum number of messages sent without notification.
+	 * Upper bound for the test parameter.
+	 */
+	BATCH_COUNT_MAX = 1024,
+	/** Count of messages in test */
+	TEST_MSG_COUNT = 1024 * 1024,
+};
+
+struct xtm_msg {
+	unsigned number;
+};
+
+/**
+ * Global pointer to xtm queue.
+ */
+static struct xtm_queue *xtm_queue;
+/** Consumer thread id */
+static pthread_t consumer_thread;
+/**
+ * Global message preallocated pointers array, we used it,
+ * because we want to test performance of xtm, not of malloc.
+ */
+static struct xtm_msg *xtm_msg_arr[TEST_MSG_COUNT];
+
+static int
+wait_for_fd(int fd)
+{
+	fd_set readset;
+	FD_ZERO(&readset);
+	FD_SET(fd, &readset);
+	return select(fd + 1, &readset, NULL, NULL, NULL);
```

Comment:
`select()` is not going to work on Mac. And I suspect on Linux it might be flaky. Because it does not support file descriptors whose number is > 1024 AFAIR. You might want to consider using `poll()`. It is almost the same but does not have this limitation. The same in the other places where `select()` is used.

Reply:
Fixed. I also modified the test to make sure that the producer and consumer thread never hang in endless waiting for each other. I run test in the loop and find, that they hangs!  There are two solutions:
1. The first is obvious: notify the consumer thread if the producer thread could not push to the queue (we abandoned it earlier).
2. Change __ATOMIC_ACQUIRE/__ATOMIC_RELEASE to __ATOMIC_SEQ_CST in file xtm/xtm_api.cc for variable is_notification_expected in push/pop/invoke functions. Also it is necessary to change  __ATOMIC_ACQUIRE/__ATOMIC_RELEASE to __ATOMIC_SEQ_CST in iterator in file xtm_spsc_queue.h. For now I do not understand why this is so.